src/kibana/plugins/discover/components/field_chooser/field_chooser.js

```diff
@@ -95,66 +94,41 @@ define(function (require) {
           _.toggleInOut($scope.columns, fieldName);
         };
 
-        var calculateFields = function (newFields) {
-          // Find the top N most popular fields
-          $scope.popularFields = _(newFields)
-          .where(function (field) {
-            return field.count > 0;
-          })
-          .sortBy('count')
-          .reverse()
-          .slice(0, config.get('fields:popularLimit'))
-          .sortBy('name')
-          .value();
-
-          // Find the top N most popular fields
-          $scope.unpopularFields = _.sortBy(_.sortBy(newFields, 'count')
-            .reverse()
-            .slice($scope.popularFields.length), 'name');
-
-          $scope.fieldTypes = _.unique(_.pluck(newFields, 'type'));
-          // push undefined so the user can clear the filter
-          $scope.fieldTypes.unshift(undefined);
-        };
-
-        $scope.$watch('fields', calculateFields);
-
-        $scope.$watch('indexPattern', function (indexPattern) {
-          $scope.fields = new IndexedArray ({
-            index: ['name'],
-            initialSet: _($scope.indexPattern.fields)
-              .sortBy('name')
-              .transform(function (fields, field) {
-                // clone the field with Object.create so that its getters
-                // and non-enumerable props are preserved
-                var clone = Object.create(field);
-                clone.display = _.contains($scope.columns, field.name);
-                fields.push(clone);
-              }, [])
-              .value()
-          });
-
-        });
+        $scope.$watchMulti([
+          '[]fieldCounts',
+          '[]columns',
```

Comment:
Turns out that rebuilding the field list is a somewhat expensive operation and could grow to be incredibly expensive for people with tons of fields.